pkg/cloud/cloud.go

```diff
@@ -1,37 +1,429 @@
+	FileCacheType            string
+	FileCacheTypeVersion     string
+	perUnitStorageThroughput int64
+}
+
+type FileCacheOptions struct {
+	CapacityGiB                          int64
+	SubnetId                             string
+	SecurityGroupIds                     []string
+	DataRepositoryAssociations           string
+	FileCacheType                        string
+	FileCacheTypeVersion                 string
+	KmsKeyId                             string
+	CopyTagsToDataRepositoryAssociations bool
+	LustreConfiguration                  []string
+	WeeklyMaintenanceStartTime           string
+	ExtraTags                            []string
+}
+
+// https://docs.aws.amazon.com/sdk-for-go/api/service/fsx/
+type FSx interface {
+	CreateFileCacheWithContext(aws.Context, *fsx.CreateFileCacheInput, ...request.Option) (*fsx.CreateFileCacheOutput, error)
+	DeleteFileCacheWithContext(aws.Context, *fsx.DeleteFileCacheInput, ...request.Option) (*fsx.DeleteFileCacheOutput, error)
+	DescribeFileCachesWithContext(aws.Context, *fsx.DescribeFileCachesInput, ...request.Option) (*fsx.DescribeFileCachesOutput, error)
+}
+
 type Cloud interface {
-	GetMetadata() MetadataService
+	CreateFileCache(ctx context.Context, volumeName string, FileCacheOptions *FileCacheOptions) (fs *FileCache, err error)
+	DeleteFileCache(ctx context.Context, FileCacheId string) (err error)
+	DescribeFileCache(ctx context.Context, FileCacheId string) (fs *FileCache, err error)
+	WaitForFileCacheAvailable(ctx context.Context, FileCacheId string) error
 }
 
 type cloud struct {
-	metadata MetadataService
+	fsx FSx
 }
 
 // NewCloud returns a new instance of AWS cloud
 // It panics if session is invalid
-func NewCloud() (Cloud, error) {
-	sess := session.Must(session.NewSession(&aws.Config{}))
-	svc := ec2metadata.New(sess)
+func NewCloud(region string) Cloud {
+	awsConfig := &aws.Config{
+		Region:                        aws.String(region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+	}
+
+	return &cloud{
+		fsx: fsx.New(session.Must(session.NewSession(awsConfig))),
+	}
+}
+
+func (c *cloud) CreateFileCache(ctx context.Context, volumeName string, fileCacheOptions *FileCacheOptions) (cache *FileCache, err error) {
+	if len(fileCacheOptions.SubnetId) == 0 {
+		return nil, fmt.Errorf("SubnetId is required")
+	}
+
+	if fileCacheOptions.DataRepositoryAssociations == "" {
+		return nil, fmt.Errorf("At least one DRA is required")
+	}
+
+	var dataRepositoryAssociations []*fsx.FileCacheDataRepositoryAssociation
+
+	dras := strings.Fields(fileCacheOptions.DataRepositoryAssociations)
+	for _, slice := range dras {
+		draConfiguration := &fsx.FileCacheDataRepositoryAssociation{}
+
+		configSlices := util.SplitUnnestedCommas(slice)
+		configMap := make(map[string]string)
+		for _, configSlice := range configSlices {
+			configSplit := strings.SplitN(configSlice, "=", 2)
+			configKey := configSplit[0]
+			configValue := configSplit[1]
+			configMap[configKey] = configValue
```

Comment:
It's a bit hard to tell what's going on here. Can we leave an example to make this code more readable?